File: leetcode/solutions/0108.ConvertSortedArrayToBinarySearchTree/sortedArrayToBST.ts
import { BinaryTreeNode } from "../../../datastructure/BinaryTreeNode.ts";

export default function sortedArrayToBST(
  nums: number[]
): BinaryTreeNode<number> | null {
  if (nums.length === 0) {
    return null;
  }

  const half = Math.floor(nums.length / 2);
  const node: BinaryTreeNode<number> = {
    val: nums[half],
    left: null,
    right: null,
  };

  node.left = sortedArrayToBST(nums.slice(0, half));
  node.right = sortedArrayToBST(nums.slice(half + 1));

  return node;
}

File: leetcode/solutions/0108.ConvertSortedArrayToBinarySearchTree/sortedArrayToBST.go
package solution

import (
	"github.com/yukisato/leetcode-util/datastructure"
)

// TreeNode is re-defined for LeetCode contest
// because they use this name as a type of binary tree node
type TreeNode = datastructure.TreeNode

func sortedArrayToBST(nums []int) *TreeNode {
	if len(nums) == 0 {
		return nil
	}

	mid := (len(nums) - 1) / 2
	root := &TreeNode{Val: nums[mid]}
	dfs(nums[:mid], root, true)
	dfs(nums[mid+1:], root, false)

	return root
}

func dfs(nums []int, tree *TreeNode, isLeft bool) {
	if len(nums) == 0 {
		return
	}

	mid := (len(nums) - 1) / 2

	if isLeft {
		tree.Left = &TreeNode{Val: nums[mid]}
		dfs(nums[:mid], tree.Left, true)
		dfs(nums[mid+1:], tree.Left, false)
	} else {
		tree.Right = &TreeNode{Val: nums[mid]}
		dfs(nums[:mid], tree.Right, true)
		dfs(nums[mid+1:], tree.Right, false)
	}
}
